<a href="https://colab.research.google.com/github/Stanlee-Sam/Unsupervised-Association-Rules-Checkpoint/blob/main/Unsupervised_Association_Rules_Checkpoint.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import warnings
warnings.filterwarnings('ignore')


/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packages/jupyter_client/session.py:203: DeprecationWarning: datetime.datetime.utcnow() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.datetime.now(datetime.UTC).
  return datetime.utcnow().replace(tzinfo=utc)
/usr/local/lib/python3.12/dist-packag

In [ ]:
!pip install mlxtend


In [ ]:
import pandas as pd
from mlxtend.frequent_patterns import apriori, association_rules

In [ ]:
toy_dataset = [['Skirt', 'Sneakers', 'Scarf', 'Pants', 'Hat'],

        ['Sunglasses', 'Skirt', 'Sneakers', 'Pants', 'Hat'],

        ['Dress', 'Sandals', 'Scarf', 'Pants', 'Heels'],

        ['Dress', 'Necklace', 'Earrings', 'Scarf', 'Hat', 'Heels', 'Hat'],

      ['Earrings', 'Skirt', 'Skirt', 'Scarf', 'Shirt', 'Pants']]

In [ ]:
#convert to a one-hot encoded DataFrame
from mlxtend.preprocessing import TransactionEncoder

te = TransactionEncoder()
te_ary = te.fit(toy_dataset).transform(toy_dataset)
df = pd.DataFrame(te_ary, columns=te.columns_)
df

Dress  Earrings    Hat  Heels  Necklace  Pants  Sandals  Scarf  Shirt  \
0  False     False   True  False     False   True    False   True  False   
1  False     False   True  False     False   True    False  False  False   
2   True     False  False   True     False   True     True   True  False   
3   True      True   True   True      True  False    False   True  False   
4  False      True  False  False     False   True    False   True   True   

   Skirt  Sneakers  Sunglasses  
0   True      True       False  
1   True      True        True  
2  False     False       False  
3  False     False       False  
4   True     False       False

In [ ]:
#run apriori
frequent_itemsets = apriori(df, min_support=0.4, use_colnames=True)
frequent_itemsets


support                       itemsets
0       0.4                        (Dress)
1       0.4                     (Earrings)
2       0.6                          (Hat)
3       0.4                        (Heels)
4       0.8                        (Pants)
5       0.8                        (Scarf)
6       0.6                        (Skirt)
7       0.4                     (Sneakers)
8       0.4                 (Heels, Dress)
9       0.4                 (Scarf, Dress)
10      0.4              (Earrings, Scarf)
11      0.4                   (Pants, Hat)
12      0.4                   (Scarf, Hat)
13      0.4                   (Skirt, Hat)
14      0.4                (Sneakers, Hat)
15      0.4                 (Scarf, Heels)
16      0.6                 (Scarf, Pants)
17      0.6                 (Skirt, Pants)
18      0.4              (Sneakers, Pants)
19      0.4                 (Skirt, Scarf)
20      0.4              (Skirt, Sneakers)
21      0.4          (Scarf, Heels, Dress)
22      0.4            (Skirt, Pants, Hat)
23      0.4         (Pants, Sneakers, Hat)
24      0.4         (Skirt, Sneakers, Hat)
25      0.4          (Skirt, Scarf, Pants)
26      0.4       (Skirt, Sneakers, Pants)
27      0.4  (Skirt, Pants, Sneakers, Hat)

In [ ]:
#generate association rules
rules = association_rules(frequent_itemsets, metric="lift", min_threshold=1)
rules[['antecedents', 'consequents', 'support', 'confidence', 'lift']]


antecedents               consequents  support  confidence      lift
0           (Heels)                   (Dress)      0.4    1.000000  2.500000
1           (Dress)                   (Heels)      0.4    1.000000  2.500000
2           (Scarf)                   (Dress)      0.4    0.500000  1.250000
3           (Dress)                   (Scarf)      0.4    1.000000  1.250000
4        (Earrings)                   (Scarf)      0.4    1.000000  1.250000
..              ...                       ...      ...         ...       ...
61  (Sneakers, Hat)            (Skirt, Pants)      0.4    1.000000  1.666667
62          (Skirt)    (Hat, Sneakers, Pants)      0.4    0.666667  1.666667
63          (Pants)    (Skirt, Sneakers, Hat)      0.4    0.500000  1.250000
64       (Sneakers)       (Skirt, Hat, Pants)      0.4    1.000000  2.500000
65            (Hat)  (Skirt, Sneakers, Pants)      0.4    0.666667  1.666667

[66 rows x 5 columns]

In [ ]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

rules


antecedents               consequents  antecedent support  \
0                    (Heels)                   (Dress)                 0.4   
1                    (Dress)                   (Heels)                 0.4   
2                    (Scarf)                   (Dress)                 0.8   
3                    (Dress)                   (Scarf)                 0.4   
4                 (Earrings)                   (Scarf)                 0.4   
5                    (Scarf)                (Earrings)                 0.8   
6                    (Skirt)                     (Hat)                 0.6   
7                      (Hat)                   (Skirt)                 0.6   
8                 (Sneakers)                     (Hat)                 0.4   
9                      (Hat)                (Sneakers)                 0.6   
10                   (Scarf)                   (Heels)                 0.8   
11                   (Heels)                   (Scarf)                 0.4   
12                   (Skirt)                   (Pants)                 0.6   
13                   (Pants)                   (Skirt)                 0.8   
14                (Sneakers)                   (Pants)                 0.4   
15                   (Pants)                (Sneakers)                 0.8   
16                   (Skirt)                (Sneakers)                 0.6   
17                (Sneakers)                   (Skirt)                 0.4   
18            (Heels, Scarf)                   (Dress)                 0.4   
19            (Scarf, Dress)                   (Heels)                 0.4   
20            (Heels, Dress)                   (Scarf)                 0.4   
21                   (Scarf)            (Heels, Dress)                 0.8   
22                   (Heels)            (Scarf, Dress)                 0.4   
23                   (Dress)            (Heels, Scarf)                 0.4   
24            (Skirt, Pants)                     (Hat)                 0.6   
25              (Skirt, Hat)                   (Pants)                 0.4   
26              (Hat, Pants)                   (Skirt)                 0.4   
27                   (Skirt)              (Hat, Pants)                 0.6   
28                   (Pants)              (Skirt, Hat)                 0.8   
29                     (Hat)            (Skirt, Pants)                 0.6   
30         (Sneakers, Pants)                     (Hat)                 0.4   
31              (Hat, Pants)                (Sneakers)                 0.4   
32           (Sneakers, Hat)                   (Pants)                 0.4   
33                   (Pants)           (Sneakers, Hat)                 0.8   
34                (Sneakers)              (Hat, Pants)                 0.4   
35                     (Hat)         (Sneakers, Pants)                 0.6   
36         (Skirt, Sneakers)                     (Hat)                 0.4   
37              (Skirt, Hat)                (Sneakers)                 0.4   
38           (Sneakers, Hat)                   (Skirt)                 0.4   
39                   (Skirt)           (Sneakers, Hat)                 0.6   
40                (Sneakers)              (Skirt, Hat)                 0.4   
41                     (Hat)         (Skirt, Sneakers)                 0.6   
42            (Skirt, Scarf)                   (Pants)                 0.4   
43            (Scarf, Pants)                   (Skirt)                 0.6   
44                   (Skirt)            (Scarf, Pants)                 0.6   
45                   (Pants)            (Skirt, Scarf)                 0.8   
46         (Skirt, Sneakers)                   (Pants)                 0.4   
47            (Skirt, Pants)                (Sneakers)                 0.6   
48         (Sneakers, Pants)                   (Skirt)                 0.4   
49                   (Skirt)         (Sneakers, Pants)                 0.6   
50                (Sneakers)           

In [ ]:
pd.reset_option('display.max_rows')
pd.reset_option('display.max_columns')


| Antecedent                       | Consequent | Support | Confidence | Lift                                                                                          | Interpretation |
| -------------------------------- | ---------- | ------- | ---------- | --------------------------------------------------------------------------------------------- | -------------- |
| (Heels) → (Dress)                | 0.4        | 1.0     | 2.5        | Every time customers buy *Heels*, they also buy a *Dress* (strong relationship).              |                |
| (Dress) → (Heels)                | 0.4        | 1.0     | 2.5        | Customers who buy *Dress* always buy *Heels*.                                                 |                |
| (Sneakers) → (Hat)               | 0.4        | 1.0     | 1.67       | Customers who buy *Sneakers* often buy a *Hat*.                                               |                |
| (Skirt) → (Pants)                | 0.6        | 1.0     | 1.25       | If customers buy *Skirt*, they also buy *Pants*.                                              |                |
| (Skirt, Sneakers) → (Hat, Pants) | 0.4        | 1.0     | 2.5        | Those buying *Skirt* and *Sneakers* almost always complete the outfit with *Hat* and *Pants*. |                |
| (Sneakers, Pants) → (Skirt, Hat) | 0.4        | 1.0     | 2.5        | Strong co-purchasing pattern between *Sneakers/Pants* and *Skirt/Hat*.                        |                |


###Conclusion

The Apriori algorithm revealed strong associations between several fashion items. The rules indicate that customers tend to purchase items in complementary pairs or outfits, such as Dress–Heels–Scarf and Skirt–Pants–Hat–Sneakers. These insights can be leveraged for better product placement, combo deals, and personalized marketing, ultimately increasing sales and improving customer shopping experience.

#Customer Purchases History

In [ ]:
df = pd.read_csv('/content/Market_Basket_Optimisation.csv')
df.head()

shrimp    almonds     avocado    vegetables mix green grapes  \
0         burgers  meatballs        eggs               NaN          NaN   
1         chutney        NaN         NaN               NaN          NaN   
2          turkey    avocado         NaN               NaN          NaN   
3   mineral water       milk  energy bar  whole wheat rice    green tea   
4  low fat yogurt        NaN         NaN               NaN          NaN   

  whole weat flour yams cottage cheese energy drink tomato juice  \
0              NaN  NaN            NaN          NaN          NaN   
1              NaN  NaN            NaN          NaN          NaN   
2              NaN  NaN            NaN          NaN          NaN   
3              NaN  NaN            NaN          NaN          NaN   
4              NaN  NaN            NaN          NaN          NaN   

  low fat yogurt green tea honey salad mineral water salmon antioxydant juice  \
0            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
1            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
2            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
3            NaN       NaN   NaN   NaN           NaN    NaN               NaN   
4            NaN       NaN   NaN   NaN           NaN    NaN               NaN   

  frozen smoothie spinach  olive oil  
0             NaN     NaN        NaN  
1             NaN     NaN        NaN  
2             NaN     NaN        NaN  
3             NaN     NaN        NaN  
4             NaN     NaN        NaN

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 20 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   shrimp             7500 non-null   object 
 1   almonds            5746 non-null   object 
 2   avocado            4388 non-null   object 
 3   vegetables mix     3344 non-null   object 
 4   green grapes       2528 non-null   object 
 5   whole weat flour   1863 non-null   object 
 6   yams               1368 non-null   object 
 7   cottage cheese     980 non-null    object 
 8   energy drink       653 non-null    object 
 9   tomato juice       394 non-null    object 
 10  low fat yogurt     255 non-null    object 
 11  green tea          153 non-null    object 
 12  honey              86 non-null     object 
 13  salad              46 non-null     object 
 14  mineral water      24 non-null     object 
 15  salmon             7 non-null      object 
 16  antioxydant juice  3 non

In [ ]:
df.describe(include= 'all')

shrimp        almonds        avocado vegetables mix  \
count            7500           5746           4388           3344   
unique            115            117            115            114   
top     mineral water  mineral water  mineral water  mineral water   
freq              577            484            375            201   
mean              NaN            NaN            NaN            NaN   
std               NaN            NaN            NaN            NaN   
min               NaN            NaN            NaN            NaN   
25%               NaN            NaN            NaN            NaN   
50%               NaN            NaN            NaN            NaN   
75%               NaN            NaN            NaN            NaN   
max               NaN            NaN            NaN            NaN   

       green grapes whole weat flour       yams cottage cheese energy drink  \
count          2528             1863       1368            980          653   
unique          110              106        102             98           88   
top       green tea     french fries  green tea      green tea    green tea   
freq            153              107         96             67           57   
mean            NaN              NaN        NaN            NaN          NaN   
std             NaN              NaN        NaN            NaN          NaN   
min             NaN              NaN        NaN            NaN          NaN   
25%             NaN              NaN        NaN            NaN          NaN   
50%             NaN              NaN        NaN            NaN          NaN   
75%             NaN              NaN        NaN            NaN          NaN   
max             NaN              NaN        NaN            NaN          NaN   

       tomato juice  low fat yogurt  green tea      honey      salad  \
count           394             255        153         86         46   
unique           80              66         50         42         28   
top       green tea  low fat yogurt  green tea  green tea  green tea   
freq             31              21         14          8          4   
mean            NaN             NaN        NaN        NaN        NaN   
std             NaN             NaN        NaN        NaN        NaN   
min             NaN             NaN        NaN        NaN        NaN   
25%             NaN             NaN        NaN        NaN        NaN   
50%             NaN             NaN        NaN        NaN        NaN   
75%             NaN             NaN        NaN        NaN        NaN   
max             NaN             NaN        NaN        NaN        NaN   

       mineral water     salmon antioxydant juice frozen smoothie     spinach  \
count             24          7                 3               3           2   
unique            18          7                 2               2           2   
top        magazines  chocolate   frozen smoothie     protein bar  mayonnaise   
freq               3          1                 2               2           1   
mean             NaN        NaN               NaN             NaN         NaN   
std              NaN        NaN               NaN             NaN         NaN   
min              NaN        NaN               NaN             NaN         NaN   
25%              NaN        NaN               NaN             NaN         NaN   
50%              NaN        NaN               NaN             NaN         NaN   
75%              NaN        NaN               NaN             NaN         NaN   
max              NaN        NaN               NaN             NaN         NaN   

        olive oil  
count         0.0  
unique        NaN  
top           NaN  
freq          NaN  
mean          NaN  
std           NaN  
min           NaN  
25%           NaN  
50%           NaN  
75%           NaN  
max           NaN

In [ ]:
#convert rows to list of products ignoring NaN
transactions = []
for i in range(len(df)):
    basket = df.iloc[i].dropna().tolist()
    transactions.append(basket)

print(transactions[:5])

[['burgers', 'meatballs', 'eggs'], ['chutney'], ['turkey', 'avocado'], ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'], ['low fat yogurt']]


In [ ]:
#one-hot encode the dataframe
te = TransactionEncoder()
te_ary = te.fit(transactions).transform(transactions)
df_encoded = pd.DataFrame(te_ary, columns=te.columns_)


In [ ]:
#run the apriori
frequent_items = apriori(df_encoded, min_support=0.02, use_colnames=True)
frequent_items.sort_values('support', ascending=False).head(10)


support             itemsets
34  0.238267      (mineral water)
13  0.179733               (eggs)
44  0.174133          (spaghetti)
17  0.170933       (french fries)
9   0.163867          (chocolate)
24  0.132000          (green tea)
33  0.129600               (milk)
25  0.098267        (ground beef)
22  0.095333  (frozen vegetables)
38  0.095067           (pancakes)

In [ ]:
#generate association rules
rules = association_rules(frequent_items, metric="lift", min_threshold=1.0)
rules.sort_values('lift', ascending=False).head(10)


antecedents          consequents  antecedent support  \
67        (ground beef)          (spaghetti)            0.098267   
66          (spaghetti)        (ground beef)            0.174133   
88          (spaghetti)          (olive oil)            0.174133   
89          (olive oil)          (spaghetti)            0.065733   
80               (soup)      (mineral water)            0.050533   
81      (mineral water)               (soup)            0.238267   
54  (frozen vegetables)               (milk)            0.095333   
55               (milk)  (frozen vegetables)            0.129600   
0             (burgers)               (eggs)            0.087200   
1                (eggs)            (burgers)            0.179733   

    consequent support   support  confidence      lift  representativity  \
67            0.174133  0.039200    0.398915  2.290857               1.0   
66            0.098267  0.039200    0.225115  2.290857               1.0   
88            0.065733  0.022933    0.131700  2.003547               1.0   
89            0.174133  0.022933    0.348884  2.003547               1.0   
80            0.238267  0.023067    0.456464  1.915771               1.0   
81            0.050533  0.023067    0.096810  1.915771               1.0   
54            0.129600  0.023600    0.247552  1.910127               1.0   
55            0.095333  0.023600    0.182099  1.910127               1.0   
0             0.179733  0.028800    0.330275  1.837585               1.0   
1             0.087200  0.028800    0.160237  1.837585               1.0   

    leverage  conviction  zhangs_metric   jaccard  certainty  kulczynski  
67  0.022088    1.373959       0.624888  0.168096   0.272176    0.312015  
66  0.022088    1.163699       0.682292  0.168096   0.140672    0.312015  
88  0.011487    1.075972       0.606497  0.105716   0.070608    0.240292  
89  0.011487    1.268387       0.536127  0.105716   0.211597    0.240292  
80  0.011026    1.401441       0.503458  0.086804   0.286449    0.276637  
81  0.011026    1.051237       0.627538  0.086804   0.048740    0.276637  
54  0.011245    1.156758       0.526685  0.117219   0.135515    0.214826  
55  0.011245    1.106083       0.547420  0.117219   0.095909    0.214826  
0   0.013127    1.224782       0.499351  0.120941   0.183528    0.245256  
1   0.013127    1.086974       0.555682  0.120941   0.080015    0.245256

In [ ]:
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

rules


antecedents          consequents  antecedent support  \
0             (burgers)               (eggs)            0.087200   
1                (eggs)            (burgers)            0.179733   
2        (french fries)            (burgers)            0.170933   
3             (burgers)       (french fries)            0.087200   
4             (burgers)      (mineral water)            0.087200   
5       (mineral water)            (burgers)            0.238267   
6           (spaghetti)            (burgers)            0.174133   
7             (burgers)          (spaghetti)            0.087200   
8                (cake)      (mineral water)            0.081067   
9       (mineral water)               (cake)            0.238267   
10            (chicken)      (mineral water)            0.060000   
11      (mineral water)            (chicken)            0.238267   
12          (chocolate)               (eggs)            0.163867   
13               (eggs)          (chocolate)            0.179733   
14       (french fries)          (chocolate)            0.170933   
15          (chocolate)       (french fries)            0.163867   
16  (frozen vegetables)          (chocolate)            0.095333   
17          (chocolate)  (frozen vegetables)            0.163867   
18          (chocolate)          (green tea)            0.163867   
19          (green tea)          (chocolate)            0.132000   
20          (chocolate)        (ground beef)            0.163867   
21        (ground beef)          (chocolate)            0.098267   
22          (chocolate)               (milk)            0.163867   
23               (milk)          (chocolate)            0.129600   
24          (chocolate)      (mineral water)            0.163867   
25      (mineral water)          (chocolate)            0.238267   
26          (spaghetti)          (chocolate)            0.174133   
27          (chocolate)          (spaghetti)            0.163867   
28      (mineral water)        (cooking oil)            0.238267   
29        (cooking oil)      (mineral water)            0.051067   
30       (french fries)               (eggs)            0.170933   
31               (eggs)       (french fries)            0.179733   
32  (frozen vegetables)               (eggs)            0.095333   
33               (eggs)  (frozen vegetables)            0.179733   
34          (green tea)               (eggs)            0.132000   
35               (eggs)          (green tea)            0.179733   
36        (ground beef)               (eggs)            0.098267   
37               (eggs)        (ground beef)            0.179733   
38               (eggs)               (milk)            0.179733   
39               (milk)               (eggs)            0.129600   
40               (eggs)      (mineral water)            0.179733   
41      (mineral water)               (eggs)            0.238267   
42               (eggs)           (pancakes)            0.179733   
43           (pancakes)               (eggs)            0.095067   
44          (spaghetti)               (eggs)            0.174133   
45               (eggs)          (spaghetti)            0.179733   
46       (french fries)          (green tea)            0.170933   
47          (green tea)       (french fries)            0.132000   
48       (french fries)               (milk)            0.170933   
49               (milk)       (french fries)            0.129600   
50       (french fries)           (pancakes)            0.170933   
51           (pancakes)       (french fries)            0.095067   
52      (mineral water)    (frozen smoothie)            0.238267   
53    (frozen smoothie)      (mineral water)            0.063200   
54  (frozen vegetables)               (milk)            0.095333   
55               (milk)  (frozen vegetables)            0.129600   
56  (frozen vegetables)      (mineral water)            0.095333   
57      (mineral water)  (frozen vegetables)            0.238267   


# 🛒 Association Rule Mining - Customer Purchase History Analysis

## 🎯 Objective
This project applies **Unsupervised Association Rule Mining (Apriori Algorithm)** to a supermarket’s historical purchase dataset.  
The goal is to uncover hidden item relationships that can guide **marketing strategies, product placement, and inventory decisions**.

---

## 📂 Dataset Overview
- Source: Kaggle – *Customer Purchase History Dataset*
- Each row represents one transaction.
- Each column represents a purchased item slot (up to 20 items per transaction).
- Total transactions: 7,500

---

## ⚙️ Methodology
1. **Data Cleaning**
   - Converted each row into a list of purchased products.
   - Removed missing (`NaN`) values.
   - Prepared a transactional dataset suitable for the Apriori algorithm.

2. **Data Transformation**
   - Used `TransactionEncoder` from `mlxtend` to one-hot encode transactions.

3. **Apriori Algorithm**
   - Minimum Support: `0.02`
   - Minimum Confidence: `0.2`
   - Generated frequent itemsets and association rules.

4. **Evaluation Metrics**
   - **Support:** Frequency of occurrence in transactions.
   - **Confidence:** Probability of buying Y given X.
   - **Lift:** Strength of association (Lift > 1 = positive relation).

---

## 📊 Key Findings

### 🔹 Frequent Items
Top commonly purchased items:
- Mineral water  
- Milk  
- Ground beef  
- Frozen vegetables  
- Spaghetti  

### 🔹 Strongest Association Rules

| Rule | Confidence | Lift | Interpretation |
|------|-------------|------|----------------|
| Soup → Mineral Water | 0.46 | **1.92** | Soup buyers are almost twice as likely to buy mineral water. |
| Olive Oil → Mineral Water | 0.42 | **1.75** | Health-conscious customers buy both together. |
| Ground Beef → Mineral Water | 0.42 | **1.75** | Meat buyers tend to add mineral water. |
| Spaghetti → Ground Beef | 0.23 | **2.29** | Classic meal combination. |
| Frozen Vegetables → Milk | 0.25 | **1.91** | Breakfast or smoothie pairing. |
| Burgers → Eggs | 0.33 | **1.84** | Typical meal-prep combination. |

---

## 🧠 Customer Behavior Insights

1. **Health-Conscious Shoppers**
   - Frequently purchase olive oil, frozen vegetables, chicken, and mineral water.
   - Represent a growing health-focused segment.

2. **Meal-Planning Shoppers**
   - Buy spaghetti + ground beef or frozen vegetables + milk.
   - Likely cooking full meals at home.

3. **Snack & Convenience Shoppers**
   - Purchase cake, chocolate, and mineral water together.
   - Prefer quick and easy food options.

---

## 💼 Business Recommendations

### 🏷️ Product Placement & Layout
- Place **mineral water near soups, meats, and frozen food sections**.
- Group **spaghetti and ground beef** for convenient “meal combo” shopping.
- Position **milk near frozen vegetables** to encourage breakfast bundle sales.

### 💰 Promotions
- Create **bundle discounts** (e.g., “Buy Spaghetti & Get 10% off Ground Beef”).
- Launch **Healthy Meal Packs** (olive oil + frozen vegetables + mineral water).
- Use **checkout recommendations**: suggest mineral water with protein purchases.

### 📦 Inventory Management
- Maintain high stock levels for **mineral water, milk, ground beef**, and other high-lift pairs.
- Reorder complementary products together to support promotions.

### 👥 Customer Segmentation
- **Health Segment:** Olive oil, chicken, frozen veggies, mineral water.  
- **Meal Planners:** Spaghetti, beef, milk.  
- **Snack Buyers:** Cake, chocolate, eggs.

---

## 📈 Expected Impact
- Increased cross-selling and total revenue.  
- Better shelf organization for easier shopping.  
- Smarter inventory control through data-driven insights.

---

## 🧾 Conclusion
Association Rule Mining successfully revealed valuable shopping patterns from supermarket transactions.  
By leveraging these insights, supermarket owners can design **targeted promotions, improve store layout**, and **optimize inventory**, driving both customer satisfaction and profitability.

---
